File: code_teller/q983/tyama_codeiq983.py
#!/usr/bin/python
import sys
if sys.version_info[0]>=3: raw_input=input

starting=5
bse=10
try:
	while True:
		n=int(raw_input())
		n+=starting-2
		digits=1
		expbase=1
		while True:
			x=digits*expbase*(bse-1)
			if x>n:break
			n-=x
			digits+=1
			expbase*=bse
		num=expbase+n//digits
		d=digits-1-n%digits
		for _ in range(d): num//=bse
		print(num%bse)
except EOFError:
	pass

File: code_teller/q983/tyama_codeiq983.pl
#!/usr/bin/perl
use strict;
use warnings;
#use bigint;
use feature qw(say);

my $starting=5;
my $bse=10;
while(<>){
	my $n=$_^0;
	$n+=$starting-2;
	my $digits=1;
	my $expbase=1;
	my $x;
	while(($x=$digits*$expbase*($bse-1))<=$n){
		$n-=$x;
		$digits++;
		$expbase*=$bse;
	}
	my $num=$expbase+$n/$digits;
	my $d=$digits-1-$n%$digits;
	for(;$d--;){
		$num/=$bse;
	}
	say $num%$bse;
}